<launch>
    <!-- launch turtlebot3_autorace_camera_calibration.launch in the raspberry pi before launch this file -->
    <arg name="mode" default="action" doc="mode type [calibration, action]"/>
    <group ns="camera">
        <!-- image brightness & contrast compensation of ground-projected image -->
        <node pkg="turtlebot3_autorace_camera" type="image_compensation" name="image_compensation" output="screen">
            <rosparam command="load" file="$(find autorace_host)/calibration/extrinsic_calibration/compensation.yaml"/>
            <param if="$(eval mode == 'calibration')" name="/is_extrinsic_camera_calibration_mode" value="True"/>
            <param if="$(eval mode == 'action')" name="/is_extrinsic_camera_calibration_mode" value="False"/>
            <remap from="/camera/image_input" to="/camera/image_rect_color"/>
            <remap from="/camera/image_input/compressed" to="/camera/image_rect_color/compressed"/>
            <remap from="/camera/image_output" to="/camera/image_compensated"/>
            <remap from="/camera/image_output/compressed" to="/camera/image_compensated/compressed"/>
        </node>

        <!-- image ground projection -->
        <node pkg="turtlebot3_autorace_camera" type="image_projection" name="image_projection" output="screen">
            <rosparam command="load" file="$(find autorace_host)/calibration/extrinsic_calibration/projection.yaml"/>
            <param if="$(eval mode == 'calibration')" name="/is_extrinsic_camera_calibration_mode" value="True"/>
            <param if="$(eval mode == 'action')" name="/is_extrinsic_camera_calibration_mode" value="False"/>
            <remap from="/camera/image_input" to="/camera/image_rect_color"/>
            <remap from="/camera/image_input/compressed" to="/camera/image_rect_color/compressed"/>
            <remap from="/camera/image_output" to="/camera/image_projected"/>
            <remap from="/camera/image_output/compressed" to="/camera/image_projected/compressed"/>
            <remap from="/camera/image_calib" to="/camera/image_extrinsic_calib"/>
            <remap from="/camera/image_calib/compressed" to="/camera/image_extrinsic_calib/compressed"/>
        </node>

        <!-- image brightness & contrast compensation of ground-projected image -->
        <node pkg="turtlebot3_autorace_camera" type="image_compensation" name="image_compensation_projection"
              output="screen">
            <rosparam command="load" file="$(find autorace_host)/calibration/extrinsic_calibration/compensation.yaml"/>
            <param if="$(eval mode == 'calibration')" name="/is_extrinsic_camera_calibration_mode" value="True"/>
            <param if="$(eval mode == 'action')" name="/is_extrinsic_camera_calibration_mode" value="False"/>
            <remap from="/camera/image_input" to="/camera/image_projected"/>
            <remap from="/camera/image_input/compressed" to="/camera/image_projected/compressed"/>
            <remap from="/camera/image_output" to="/camera/image_projected_compensated"/>
            <remap from="/camera/image_output/compressed" to="/camera/image_projected_compensated/compressed"/>
        </node>
    </group>

    <!-- traffic sign detection -->

    <node pkg="turtlebot3_autorace_detect" type="detect_intersection_sign" name="detect_intersection_sign"
          output="screen">

        <remap from="/detect/image_input" to="/camera/image_compensated"/>
        <remap from="/detect/image_input/compressed" to="/camera/image_compensated/compressed"/>
        <remap from="/detect/image_output" to="/detect/image_traffic_sign"/>
        <remap from="/detect/image_output/compressed" to="/detect/image_traffic_sign/compressed"/>
    </node>

    <node pkg="turtlebot3_autorace_detect" type="detect_construction_sign" name="detect_construction_sign"
          output="screen">

        <remap from="/detect/image_input" to="/camera/image_compensated"/>
        <remap from="/detect/image_input/compressed" to="/camera/image_compensated/compressed"/>
        <remap from="/detect/image_output" to="/detect/image_traffic_sign"/>
        <remap from="/detect/image_output/compressed" to="/detect/image_traffic_sign/compressed"/>
    </node>

    <node pkg="turtlebot3_autorace_detect" type="detect_parking_sign" name="detect_parking_sign"
          output="screen">

        <remap from="/detect/image_input" to="/camera/image_compensated"/>
        <remap from="/detect/image_input/compressed" to="/camera/image_compensated/compressed"/>
        <remap from="/detect/image_output" to="/detect/image_traffic_sign"/>
        <remap from="/detect/image_output/compressed" to="/detect/image_traffic_sign/compressed"/>
    </node>

    <node pkg="turtlebot3_autorace_detect" type="detect_level_crossing_sign" name="detect_level_crossing_sign"
          output="screen">

        <remap from="/detect/image_input" to="/camera/image_compensated"/>
        <remap from="/detect/image_input/compressed" to="/camera/image_compensated/compressed"/>
        <remap from="/detect/image_output" to="/detect/image_traffic_sign"/>
        <remap from="/detect/image_output/compressed" to="/detect/image_traffic_sign/compressed"/>
    </node>

    <node pkg="turtlebot3_autorace_detect" type="detect_tunnel_sign" name="detect_tunnel_sign"
          output="screen">

        <remap from="/detect/image_input" to="/camera/image_compensated"/>
        <remap from="/detect/image_input/compressed" to="/camera/image_compensated/compressed"/>
        <remap from="/detect/image_output" to="/detect/image_traffic_sign"/>
        <remap from="/detect/image_output/compressed" to="/detect/image_traffic_sign/compressed"/>
    </node>

</launch>